### DISCUSSION TITLE
1.6.0 Custom input with typescript break build

### DISCUSSION THREAD

**Discussion by Xeyos88** [formkit/formkit#1258] [General]
Hi,
I have some custom inputs in my configuration e.g:
```
const listSeparator = createInput(ListSeparatorFormKit)
....
const config: DefaultConfigOptions = {
  plugins: [changeLabelPlugin()],
  config: {
    classes: generateClasses({
     .....
    }),
  },
  inputs: {
    listSeparator,
  },
}
```
And use it in this way:
```
 <FormKit
       type="listSeparator"
       name="paths"
       label="Input paths"
      placeholder="Set input paths"
      validation="required"
     />
```
I use typescript and until the version 1.5.9 I didn't have any problems.

With the versione 1.6.0 I have this error:

> error TS2322: Type '"listSeparator"' is not assignable to type '"number" | "range" | "month" | "time" | "search" | "list" | "button" | "form" | "meta" | "select" | "textarea" | "text" | "hidden" | "color" | "checkbox" | "date" | "email" | "file" | ... 8 more ... | undefined'.
> 
> 321               type="listSeparator"

I looked at the documentation but there is nothing different

**Comment by dominikklein** [comment]
Did you already check the changelog?

https://formkit.com/changelog

There is this note:

```
Custom Input Types

The type argument of the FormKit component now requires a matching type in the FormKitInputProps type for TypeScript support. If you are using custom inputs and have supplied a module augmentation for FormKitInputProps, you will need to [update the FormKitInputProps type to include your types](https://formkit.com/essentials/custom-inputs).
``` 

https://formkit.com/essentials/custom-inputs

**Reply by Xeyos88** [reply]
Ok in quick check I missed the part about the extension of  FormKitInputProps. 
I will try thx

**Reply by Xeyos88** [reply]
The solution works but with a doubt:
I have 4 custom types and have currently increased the definition for only one of them, but the build goes through and everything works.
I don't understand why it is not necessary to define all 4 types.
I also tried to put as a key a value that is not among the 4 types I defined and it still works.

**Reply by Finrod927** [reply]
Hi
I also had the same issue and also missed the info in the changelog on the site (I checked the changelog in the repo, nothing there about that).
I also have some doubt about that augmentation, once I created a d.ts file with `declare module '@formkit/inputs'` the error was gone, and no matter what I put in the module, the error does not show up.

But it breaks the autocompletion, even with correct keys in the `FormKitInputProps`, the autompletion of the FormKit component is broken: props are not correctly suggested, and the `type` props does not suggest the FormKit input types.

@Xeyos88, did you figure it out?

**Reply by justin-schroeder** [reply]
Can you provide a minimal reproduction of this in a new repository?

**Reply by Finrod927** [reply]
I just fixed it,
I'm not sure it's the correct way to do it (I'm not very familiar with module augmentation), but declaring the module augmentation in a `.ts` file instead of a `.d.ts` file seems to work.

The error is now shown on custom types I did not declared, and the auto-completion is working and suggesting custom types I declared in the module.

**Reply by remisdavid** [reply]
@Finrod927 I am losing my mind over this. Can you please post your code that solved this? 🙏

**Reply by Finrod927** [reply]
@remisdavid sure!

I have a wrapper component to be able to use FormKit with Vuetify components. I defined these custom components in a `formkit.ts`, which also contains this module augmentation:

```ts
/**
 * Augmenting FormKit type definition with custom inputs
 * https://formkit.com/essentials/custom-inputs#adding-prop-types
 */
declare module '@formkit/inputs' {
  interface FormKitInputProps<Props extends FormKitInputs<Props>> {
    // This key and the `type` should match
    VTextarea: { type: 'VTextarea' }
    VTextField: { type: 'VTextField' }
  }

  // Vuetify slots
  interface VuetifySlots {
    append: VInputAppendSlot
    'append-inner': VSelectAppendInnerSlot
  }

  interface FormKitInputSlots<Props extends FormKitInputs<Props>> {
    VTextarea: FormKitTextSlots<Props> & VuetifySlots
    VTextField: FormKitTextSlots<Props> & VuetifySlots
  }
}
```

This allow me to use a custom type in the Formkit component:
```vue
    <FormKit
      id="descriptionField"
      name="description"
      type="VTextarea"
      label="Description"
    />

```

The only downside I did not figured it out yet, is the slots management: I need to re-declare all slots I need in the custom component and in the module augmentation.

I hope this will help 😊

**Reply by remisdavid** [reply]
@Finrod927 This really did solve it. Thanks!

But there is still one thing unsolved for me. Did you managed to solve the issue with autocompletion and linting for additional props?

In your case making such input type:
```ts
  interface FormKitInputProps<Props extends FormKitInputs<Props>> {
    VTextField: { type: 'VTextField', superImportantProp: string }
  }
```

and when using the type and missing the prop like such, there will be error:
```vue
    <FormKit
      id="descriptionField"
      name="description"
      type="VTextField"
      label="Description"
      /*superImportantProp missing here*/
    />
```

The docs suggest it here, but I haven't been able to make it work :/ https://formkit.com/essentials/custom-inputs#adding-prop-types

**Reply by Finrod927** [reply]
Indeed, I do not have autocompletion nor linting on additional props.
But as I only have the "type" (which VSCode has in his default prop list) and another props that I barely used, I didn't realized it wasn't working 😅
